File: docs/validation/substrate_math_performance/test-harness/swift/Sources/Harness/Primitives/MerkleCommitmentSpike.swift
// MerkleCommitmentSpike.swift
//
// The NT-P0 bakeoff's Merkle/commitment byte contract, RECOVERED from git
// history into the harness as harness-local support code.
//
// Provenance: this contract arrived with the bakeoff (d848a1950,
// "feat(nt-p0): Merkle/commitment contract") as
// SubstrateKernel/MerkleCommitment.swift and was correctly deleted from
// SubstrateKernel (a7ee72763) because the PRODUCTION Merkle implementation
// is SubstrateLib/MerkleHash.swift (NT-F2) — but that removal never chased
// the two harness consumers (MerkleCommitmentPrimitive, NTP0Bakeoff), which
// left this package uncompilable. The Rust harness never had the problem:
// its copy lives inside the harness (rust/src/primitives/merkle_commitment.rs).
// This file mirrors that layout for Swift: the bakeoff contract belongs to
// the bakeoff tooling, not to the substrate.
//
// Changes from the deleted spike (API drift since d848a1950):
//   • MerkleRoot(unchecked:) / ContentHash(unchecked:) → init(bytes:)
//   • MerkleRoot(contentHash:) → MerkleRoot(bytes: hash.bytes)
//   • the spike-era `KeyedCommitment` type (also deleted) is replaced by the
//     local `MerkleKeyedCommitment` below, shaped to what the harness reads
//     (.wireBytes, .keyVersion as UInt32)
//   • spike-era `.wireBytes` spellings on MerkleRoot/ContentHash are kept
//     working via the compatibility extensions at the bottom so the
//     consumer files stay verbatim to their bakeoff-era form.
//
// Hashing uses the in-repo SHA256 primitive; keyed commitments reuse
// GrantHKDF.hmac so there is only one HMAC-SHA256 construction in play.

import Foundation
import SubstrateKernel
import SubstrateTypes

public struct MerkleVectorPayload: Sendable, Hashable {
    public let modelID: String
    public let vectorIndex: UInt32
    public let values: [Float32]

    public init(modelID: String, vectorIndex: UInt32, values: [Float32]) {
        self.modelID = modelID
        self.vectorIndex = vectorIndex
        self.values = values
    }
}

public struct MerkleChild: Sendable, Hashable {
    public let childID: UUID
    public let root: MerkleRoot

    public init(childID: UUID, root: MerkleRoot) {
        self.childID = childID
        self.root = root
    }
}

/// Harness-local stand-in for the spike-era `KeyedCommitment` type (deleted
/// with the spike). Carries exactly what the harness consumers read: the
/// 32-byte HMAC (`wireBytes`) and the producing key version.
public struct MerkleKeyedCommitment: Sendable, Hashable {
    public let wireBytes: [UInt8]
    public let keyVersion: UInt32

    public init(wireBytes: [UInt8], keyVersion: UInt32) {
        precondition(wireBytes.count == 32,
                     "HMAC-SHA256 output must be exactly 32 bytes")
        self.wireBytes = wireBytes
        self.keyVersion = keyVersion
    }
}

public enum MerkleCommitment {
    public enum DomainTag {
        public static let leaf: UInt8 = 0x00
        public static let interior: UInt8 = 0x01
        public static let tombstone: UInt8 = 0x02
        public static let keyedCommitment: UInt8 = 0x03
    }

    /// Named empty-subtree payload. The empty root is SHA256 over
    /// the bare INTERIOR domain tag (a node with zero children).
    public static let emptyRootPayload: [UInt8] = [DomainTag.interior]

    public static let emptyRoot = MerkleRoot(
        bytes: SHA256.hash(emptyRootPayload)
    )

    /// Canonical leaf bytes:
    ///
    /// 1. leaf domain tag
    /// 2. drawer UUID raw 16-byte big-endian form
    /// 3. u64 big-endian byte length + NFC UTF-8 content bytes
    /// 4. u32 big-endian vector record count
    /// 5. vector records sorted by (modelID UTF-8 bytes, vectorIndex)
    ///
    /// Each vector record encodes modelID length+bytes, vectorIndex,
    /// value count, then each Float32 as IEEE-754 little-endian bytes.
    public static func canonicalLeafPayload(
        drawerID: UUID,
        content: String,
        vectors: [MerkleVectorPayload]
    ) -> [UInt8] {
        let nfc = content.precomposedStringWithCanonicalMapping
        return canonicalLeafPayload(
            drawerIDBytes: uuidBytes(drawerID),
            contentNFCUTF8: Array(nfc.utf8),
            vectors: vectors
        )
    }

    public static func canonicalLeafPayload(
        drawerIDBytes: [UInt8],
        contentNFCUTF8: [UInt8],
        vectors: [MerkleVectorPayload]
    ) -> [UInt8] {
        precondition(drawerIDBytes.count == 16, "drawerIDBytes must be 16 bytes")
        precondition(vectors.count <= Int(UInt32.max),
                     "too many vector records for u32 length prefix")

        var bytes = [UInt8]()
        bytes.reserveCapacity(1 + 16 + 8 + contentNFCUTF8.count + 4)
        bytes.append(DomainTag.leaf)
        bytes.append(contentsOf: drawerIDBytes)
        appendBigEndian(UInt64(contentNFCUTF8.count), to: &bytes)
        bytes.append(contentsOf: contentNFCUTF8)

        let ordered = vectors.sorted { lhs, rhs in
            let l = Array(lhs.modelID.utf8)
            let r = Array(rhs.modelID.utf8)
            if l != r { return lexicographicallyPrecedes(l, r) }
            return lhs.vectorIndex < rhs.vectorIndex
        }
        appendBigEndian(UInt32(ordered.count), to: &bytes)
        for vector in ordered {
            let modelBytes = Array(vector.modelID.utf8)
            precondition(modelBytes.count <= Int(UInt32.max),
                         "modelID too large for u32 length prefix")
            precondition(vector.values.count <= Int(UInt32.max),
                         "vector too large for u32 value count")
            appendBigEndian(UInt32(modelBytes.count), to: &bytes)
            bytes.append(contentsOf: modelBytes)
            appendBigEndian(vector.vectorIndex, to: &bytes)
            appendBigEndian(UInt32(vector.values.count), to: &bytes)
            for value in vector.values {
                appendLittleEndian(value.bitPattern, to: &bytes)
            }
        }
        return bytes
    }

    public static func leafHash(
        drawerID: UUID,
        content: String,
        vectors: [MerkleVectorPayload] = []
    ) -> ContentHash {
        let payload = canonicalLeafPayload(drawerID: drawerID, content: content, vectors: vectors)
        return ContentHash(bytes: SHA256.hash(payload))
    }

    public static func leafRoot(
        drawerID: UUID,
        content: String,
        vectors: [MerkleVectorPayload] = []
    ) -> MerkleRoot {
        MerkleRoot(bytes: leafHash(drawerID: drawerID, content: content, vectors: vectors).bytes)
    }

    public static func hashLeafPayload(_ payload: [UInt8]) -> ContentHash {
        precondition(payload.first == DomainTag.leaf,
                     "leaf payload must start with the leaf domain tag")
        return ContentHash(bytes: SHA256.hash(payload))
    }

    public static func rootForLeafPayload(_ payload: [UInt8]) -> MerkleRoot {
        MerkleRoot(bytes: hashLeafPayload(payload).bytes)
    }

    /// Canonical interior bytes: INTERIOR tag plus child roots in
    /// ascending child-id byte order.
    public static func canonicalInteriorPayload(children: [MerkleChild]) -> [UInt8] {
        var bytes = [UInt8]()
        bytes.reserveCapacity(1 + children.count * 32)
        bytes.append(DomainTag.interior)
        let ordered = children.sorted {
            lexicographicallyPrecedes(uuidBytes($0.childID), uuidBytes($1.childID))
        }
        for child in ordered {
            bytes.append(contentsOf: child.root.bytes)
        }
        return bytes
    }

    public static func interiorRoot(children: [MerkleChild]) -> MerkleRoot {
        MerkleRoot(bytes: SHA256.hash(canonicalInteriorPayload(children: children)))
    }

    public static func canonicalTombstonePayload(drawerID: UUID) -> [UInt8] {
        var bytes = [UInt8]()
        bytes.reserveCapacity(17)
        bytes.append(DomainTag.tombstone)
        bytes.append(contentsOf: uuidBytes(drawerID))
        return bytes
    }

    public static func tombstoneHash(drawerID: UUID) -> ContentHash {
        ContentHash(bytes: SHA256.hash(canonicalTombstonePayload(drawerID: drawerID)))
    }

    public static func tombstoneRoot(drawerID: UUID) -> MerkleRoot {
        MerkleRoot(bytes: tombstoneHash(drawerID: drawerID).bytes)
    }

    public static func keyedCommitment(
        drawerID: UUID,
        content: String,
        vectors: [MerkleVectorPayload] = [],
        key: [UInt8],
        keyVersion: UInt32
    ) -> MerkleKeyedCommitment {
        let payload = canonicalLeafPayload(drawerID: drawerID, content: content, vectors: vectors)
        return keyedCommitment(forCanonicalLeafPayload: payload, key: key, keyVersion: keyVersion)
    }

    public static func keyedCommitment(
        forCanonicalLeafPayload payload: [UInt8],
        key: [UInt8],
        keyVersion: UInt32
    ) -> MerkleKeyedCommitment {
        precondition(payload.first == DomainTag.leaf,
                     "keyed commitment input must be a canonical leaf payload")
        var committed = [UInt8]()
        committed.reserveCapacity(1 + payload.count)
        committed.append(DomainTag.keyedCommitment)
        committed.append(contentsOf: payload)
        let mac = GrantHKDF.hmac(key: key, data: committed)
        return MerkleKeyedCommitment(wireBytes: mac, keyVersion: keyVersion)
    }

    // MARK: - Encoding helpers

    public static func uuidBytes(_ uuid: UUID) -> [UInt8] {
        let u = uuid.uuid
        return [
            u.0, u.1, u.2, u.3,
            u.4, u.5, u.6, u.7,
            u.8, u.9, u.10, u.11,
            u.12, u.13, u.14, u.15,
        ]
    }

    private static func appendBigEndian(_ value: UInt64, to bytes: inout [UInt8]) {
        for shift in stride(from: 56, through: 0, by: -8) {
            bytes.append(UInt8((value >> shift) & 0xFF))
        }
    }

    private static func appendBigEndian(_ value: UInt32, to bytes: inout [UInt8]) {
        for shift in stride(from: 24, through: 0, by: -8) {
            bytes.append(UInt8((value >> shift) & 0xFF))
        }
    }

    private static func appendLittleEndian(_ value: UInt32, to bytes: inout [UInt8]) {
        for shift in stride(from: 0, through: 24, by: 8) {
            bytes.append(UInt8((value >> shift) & 0xFF))
        }
    }

    private static func lexicographicallyPrecedes(_ lhs: [UInt8], _ rhs: [UInt8]) -> Bool {
        for i in 0..<min(lhs.count, rhs.count) {
            if lhs[i] != rhs[i] { return lhs[i] < rhs[i] }
        }
        return lhs.count < rhs.count
    }
}

// MARK: - Spike-era spelling compatibility

/// The bakeoff-era consumer files (MerkleCommitmentPrimitive, NTP0Bakeoff)
/// call `.wireBytes` and `init(unchecked:)` on MerkleRoot and ContentHash;
/// the production types (post-NT-F2) spell them `.bytes` / `init(bytes:)`.
/// Harness-local aliases keep those files verbatim to their recorded
/// bakeoff form; the `unchecked:` spelling forwards to the validating
/// production init, so nothing is actually unchecked anymore.
extension MerkleRoot {
    public var wireBytes: [UInt8] { bytes }
    public init(unchecked bytes: [UInt8]) { self.init(bytes: bytes) }
}

extension ContentHash {
    public var wireBytes: [UInt8] { bytes }
    public init(unchecked bytes: [UInt8]) { self.init(bytes: bytes) }
}

File: docs/validation/substrate_math_performance/test-harness/rust/src/primitives/merkle_commitment_spike.rs
// src/primitives/merkle_commitment_spike.rs
//
// The NT-P0 bakeoff's Merkle/commitment byte contract, vendored into the
// harness as harness-local support code. Rust twin of the Swift recovery
// (MerkleCommitmentSpike.swift, 296a4c74).
//
// Provenance: this contract arrived with the bakeoff as a SubstrateKernel
// spike and was correctly deleted from the kernel (the PRODUCTION Merkle
// implementation is SubstrateLib merkle_hash.rs / MerkleHash.swift, NT-F2)
// — but the CE Rust harness primitive (merkle_commitment.rs) arrived via
// the EE 06-29 SHARED mirror still importing the EE kernel's
// `substrate_kernel::merkle_commitment`, a module that never existed in
// the CE kernel, leaving this package uncompilable. The Swift harness got
// its vendored copy in 296a4c74; this module completes the recovery for
// Rust: the bakeoff contract belongs to the bakeoff tooling, not to the
// substrate.
//
// The byte contract is identical to production merkle_hash (leaf layout,
// domain tags, interior sort, empty root) — the bakeoff is where that
// contract was ratified — so the checked-in vectors/merkle_commitment.json
// remain valid unchanged.
//
// Hashing uses the in-repo substrate_kernel::sha256; keyed commitments
// reuse substrate_kernel::hkdf::hmac so there is only one HMAC-SHA256
// construction in play — mirroring the Swift spike's SHA256/GrantHKDF use.

use substrate_kernel::{hkdf, sha256};
use substrate_types::content_hash::ContentHash;
use substrate_types::merkle_domain::MerkleDomain;
use substrate_types::MerkleRoot;

/// Borrowed vector record for the canonical leaf payload. Mirrors the
/// Swift spike's `MerkleVectorPayload` (borrowed here because the harness
/// builds records over slices it already owns).
pub struct MerkleVectorPayload<'a> {
    pub model_id: &'a str,
    pub vector_index: u32,
    pub values: &'a [f32],
}

impl<'a> MerkleVectorPayload<'a> {
    pub fn new(model_id: &'a str, vector_index: u32, values: &'a [f32]) -> Self {
        Self { model_id, vector_index, values }
    }
}

/// One child of an interior node: raw 16-byte big-endian UUID + its root.
pub struct MerkleChild {
    pub child_id: [u8; 16],
    pub root: MerkleRoot,
}

impl MerkleChild {
    pub fn new(child_id: [u8; 16], root: MerkleRoot) -> Self {
        Self { child_id, root }
    }
}

/// Harness-local stand-in for the spike-era `KeyedCommitment` type.
/// Carries exactly what the harness consumers read: the 32-byte HMAC
/// (`wire_bytes`/`as_bytes`) and the producing key version. Mirrors the
/// Swift spike's `MerkleKeyedCommitment`.
pub struct MerkleKeyedCommitment {
    wire: [u8; 32],
    pub key_version: u32,
}

impl MerkleKeyedCommitment {
    pub fn wire_bytes(&self) -> [u8; 32] {
        self.wire
    }

    pub fn as_bytes(&self) -> &[u8; 32] {
        &self.wire
    }
}

/// Canonical leaf bytes (bakeoff contract, byte-identical to production
/// canonical_leaf_bytes):
///
/// 1. leaf domain tag (0x00)
/// 2. drawer UUID raw 16-byte big-endian form
/// 3. u64 big-endian byte length + content bytes (the caller supplies
///    already-normalized UTF-8; no NFC pass happens here, matching the
///    Swift spike's byte-level overload)
/// 4. u32 big-endian vector record count
/// 5. vector records sorted by (model_id UTF-8 bytes, vector_index),
///    each as u32 BE model_id length | model_id bytes | u32 BE
///    vector_index | u32 BE value count | IEEE-754 LE float bytes
pub fn canonical_leaf_payload(
    drawer_id: [u8; 16],
    content: &[u8],
    vectors: &[MerkleVectorPayload<'_>],
) -> Vec<u8> {
    let mut bytes = Vec::with_capacity(1 + 16 + 8 + content.len() + 4);
    bytes.push(MerkleDomain::LEAF);
    bytes.extend_from_slice(&drawer_id);
    bytes.extend_from_slice(&(content.len() as u64).to_be_bytes());
    bytes.extend_from_slice(content);

    let mut order: Vec<usize> = (0..vectors.len()).collect();
    order.sort_by(|&a, &b| {
        vectors[a]
            .model_id
            .as_bytes()
            .cmp(vectors[b].model_id.as_bytes())
            .then(vectors[a].vector_index.cmp(&vectors[b].vector_index))
    });
    bytes.extend_from_slice(&(vectors.len() as u32).to_be_bytes());
    for i in order {
        let v = &vectors[i];
        let model_bytes = v.model_id.as_bytes();
        bytes.extend_from_slice(&(model_bytes.len() as u32).to_be_bytes());
        bytes.extend_from_slice(model_bytes);
        bytes.extend_from_slice(&v.vector_index.to_be_bytes());
        bytes.extend_from_slice(&(v.values.len() as u32).to_be_bytes());
        for value in v.values {
            bytes.extend_from_slice(&value.to_bits().to_le_bytes());
        }
    }
    bytes
}

/// SHA-256 over a canonical leaf payload. The payload must carry the
/// leaf domain tag (it was built by `canonical_leaf_payload`).
pub fn hash_leaf_payload(payload: &[u8]) -> ContentHash {
    assert_eq!(
        payload.first(),
        Some(&MerkleDomain::LEAF),
        "leaf payload must start with the leaf domain tag"
    );
    ContentHash::new(sha256::hash(payload))
}

/// Interior root: INTERIOR tag plus child roots in ascending child-id
/// byte order.
pub fn interior_root(children: &[MerkleChild]) -> MerkleRoot {
    let mut order: Vec<usize> = (0..children.len()).collect();
    order.sort_by(|&a, &b| children[a].child_id.cmp(&children[b].child_id));

    let mut payload = Vec::with_capacity(1 + children.len() * 32);
    payload.push(MerkleDomain::INTERIOR);
    for i in order {
        payload.extend_from_slice(children[i].root.bytes());
    }
    MerkleRoot::new(sha256::hash(&payload))
}

/// Tombstone hash: TOMBSTONE tag (0x02) + drawer id 16B BE.
pub fn tombstone_hash(drawer_id: [u8; 16]) -> ContentHash {
    let mut payload = Vec::with_capacity(17);
    payload.push(MerkleDomain::TOMBSTONE);
    payload.extend_from_slice(&drawer_id);
    ContentHash::new(sha256::hash(&payload))
}

/// Empty-subtree root: SHA-256 over the bare INTERIOR domain tag. Equal
/// by construction to `MerkleRoot::EMPTY` (whose literal a kernel bridge
/// test pins to this same computation); computed here so the harness
/// validates the derivation, not the literal.
pub fn empty_root() -> MerkleRoot {
    MerkleRoot::new(sha256::hash(&[MerkleDomain::INTERIOR]))
}

/// HMAC-SHA256 commitment over a canonical leaf payload with the
/// COMMITMENT domain tag (0x03) prepended.
pub fn keyed_commitment_for_canonical_leaf_payload(
    payload: &[u8],
    key: &[u8],
    key_version: u32,
) -> MerkleKeyedCommitment {
    assert_eq!(
        payload.first(),
        Some(&MerkleDomain::LEAF),
        "keyed commitment input must be a canonical leaf payload"
    );
    let mut committed = Vec::with_capacity(1 + payload.len());
    committed.push(MerkleDomain::COMMITMENT);
    committed.extend_from_slice(payload);
    MerkleKeyedCommitment { wire: hkdf::hmac(key, &committed), key_version }
}

/// Spike-era spelling compatibility: the bakeoff-era consumer file
/// (merkle_commitment.rs) calls `.wire_bytes()` / `.as_bytes()` on
/// MerkleRoot and ContentHash; the production types (post-NT-F2) spell
/// the accessor `.bytes()`. This trait keeps the consumer verbatim to
/// its recorded bakeoff form — the Rust analogue of the Swift spike's
/// `extension MerkleRoot { var wireBytes … }`.
pub trait SpikeWireBytes {
    fn wire_bytes(&self) -> [u8; 32];
    fn as_bytes(&self) -> &[u8; 32];
}

impl SpikeWireBytes for MerkleRoot {
    fn wire_bytes(&self) -> [u8; 32] {
        *self.bytes()
    }
    fn as_bytes(&self) -> &[u8; 32] {
        self.bytes()
    }
}

impl SpikeWireBytes for ContentHash {
    fn wire_bytes(&self) -> [u8; 32] {
        *self.bytes()
    }
    fn as_bytes(&self) -> &[u8; 32] {
        self.bytes()
    }
}
